### PR TITLE
[BUU] Activate admin_style_v3 for most system specs

### PR DESCRIPTION
**Pull request by filipefurtad0**
#### What? Why?

!! Not to be merged just yet, as it activates BUU by default !!

- Required for https://github.com/openfoodfoundation/openfoodnetwork/issues/12627

It does not yet fully address feature parity, I've disconnected the issue [#11845](https://github.com/openfoodfoundation/openfoodnetwork/issues/11845)

- Enables the BUU toggle on `base_spec_helper.rb`;
- Fixes case on text assertions
- Fixes failing tests due to overlapping elements; a helper was created to deal with this

In what #11845 concerns, this is the current state (this will be updated as progressed):
```

As an Administrator
  I want to be able to manage products in bulk
  listing products
    displays a list of products - **DONE**
    displays a message when number of products is zero - **DONE**
    displays a select box for suppliers, with the appropriate supplier selected - **PENDING (should be fixed in #11932)**
    displays an on hand count in a span for each product - **DONE**
    displays 'on demand' for any variant that is available on demand - **DONE**
    displays a select box for the unit of measure for the product's variants - **PENDING**
    displays a text field for the item name when unit is set to 'Items' - **PENDING**

```
#### What should we test?

- System tests should display the BUU design on all admin pages
- Green build

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

#### Dependencies
This can be done once we have completed all BUU1 features, including:
* #12473 
* #12331 
* #12398 
* #11987 
* #11055 
* #12482 
* #12362

### CODE REVIEW

**Comment by filipefurtad0** [comment]
Ok, some `/admin` specs are failing, moving to in Dev.

**Comment by dacook** [comment]
Thanks Filipe! It looks like probably all of these failures are due to a style change in capitalisation, which results in text no longer matching. Eg:

```
  expected to find text "DISTRIBUTOR ..." in "Distributor..."
```

I'm not sure of the most efficient way to solve this, but I've often thought that these expectations shouldn't be case-sensitive (they still convey the same meaning to the user), so I would argue that we change that. I'll raise in #core-devs.

**Comment by dacook** [comment]
I [discussed with Maikel](https://openfoodnetwork.slack.com/archives/GK2T38QPJ/p1696977666895229), and agreed it's worth updating the text in the specs to be more readable and better match what's on the screen.

We also agreed that it's helpful to optionally add a matcher like [`have_content_i`](https://stackoverflow.com/a/44575326/421243) to better handle this in the future. 

@filipefurtad0 would you like to try this, or perhaps hand over for a `good-first-issue`?

**Comment by filipefurtad0** [comment]
Happy to have a go at this :+1: 
Thanks @dacook

File: spec/system/admin/subscriptions/crud_spec.rb
Comment (dacook) [inline spec/system/admin/subscriptions/crud_spec.rb:131]: 🧐

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks Filipe

Since you started this, the `/admin/products` URL now points to the new screen, so the old `bulk_product_update_spec` now fails (it's loading the new one instead of the old one). I think you should be able to fix this by disabling the feature toggle on that test.

File: spec/system/admin/subscriptions/crud_spec.rb
Comment (filipefurtad0) [reply spec/system/admin/subscriptions/crud_spec.rb:131]: Oops!

**Comment by filipefurtad0** [comment]
Hum, indeed there are some failing tests now. I've left this one forgotten in In Dev, but now of course the code has changed (after rebasing).

Either way, even after fixing, I guess we'll only be able to merge this PR once we're ready to release the feature, right? So, maybe we should hold this one for a while... And update everything once we're ready to release BUU.

What do you think @dacook ?

**Comment by dacook** [comment]
🤔 Hmm. 

* We want to continue supporting the old (current) admin styles, so need to ensure it's being tested
* We want to also test with the new styles, in case any new problems come up

We could run the specs twice, for both old and new. 
But I don't think it's necessary. Apart from the bulk products screen, the only changes are styling (colours, padding, fonts, etc). 

So I agree, let's keep this PR in dev, and wait until we get closer to releasing admin_style_v3. We can occasionally refresh this branch when needed until then.

**Comment by filipefurtad0** [comment]
> So I agree, let's keep this PR in dev, and wait until we get closer to releasing admin_style_v3. We can occasionally refresh this branch when needed until then.

Ok, great - thanks. I've updated it with a few changes so it's green again.

**Comment by filipefurtad0** [comment]
Ok, I've fixed conflicts, rebased, updated a test - it's green again.

It seems to me the challenge now, is to migrate the relevant [tests from the legacy products page](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/bulk_product_update_spec.rb) into the spec for [new product page](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/products_v3/products_spec.rb); After that, we can delete the legacy spec file.

Does this sound like a good plan @dacook?

**Comment by dacook** [comment]
Awesome, thanks Filipe! I've created an issue for this: 
- https://github.com/openfoodfoundation/openfoodnetwork/issues/11845

I think we shouldn't delete the legacy spec until we remove the feature toggle.

File: spec/support/tom_select_helper.rb
Comment (dacook) [inline spec/support/tom_select_helper.rb:11]: Here `all` is potentially selecting multiple options (eg a value of "Distributor" could select "Distributor1" and "Distributor2"), then clicking on the first one. 
I think what we want is to use `find`, and I'd suggest against using Regexp if it's not needed. We would be looking for an exact match.

File: spec/support/tom_select_helper.rb
Comment (dacook) [inline spec/support/tom_select_helper.rb:8]: I'm wondering if this is necessary. Doesn't capybara automatically scroll and find the desired element without the need for searching?

If needed, I wonder does the simpler `fill_in '.ts-control' with: value`  work?

File: spec/support/tom_select_helper.rb
Comment (rioug) [inline spec/support/tom_select_helper.rb:3]: There are already some helper for tom select here :
https://github.com/openfoodfoundation/openfoodnetwork/blob/dd502f0883d8099edec197d3368eb5d9420669ea/spec/support/request/web_helper.rb#L93
But like the idea of moving them to their own helper.

**Review by dacook — COMMENTED** [review]
Looks ok to me. I have some ideas on simplifying the code, but it's possible that this code was necessary to get it working.

I'd love to see official helpers for this stuff, but sadly they [don't exist yet](https://github.com/orchidjs/tom-select/discussions/71#discussioncomment-5844625). Maybe one day we'll be able to contribute them :D

File: spec/support/tom_select_helper.rb
Comment (rioug) [reply spec/support/tom_select_helper.rb:8]: It does currently, but once we move to the latest version of tom select using `fill_in '.ts-control' with: value` will break.

File: spec/support/tom_select_helper.rb
Comment (filipefurtad0) [reply spec/support/tom_select_helper.rb:3]: Ahh, I was unaware of those existing helpers... I'll move them to this file. The build should actually stay green, as we're calling the new file from `base_spec_helper`...

File: spec/support/tom_select_helper.rb
Comment (filipefurtad0) [reply spec/support/tom_select_helper.rb:8]: So, I guess we leave this one as is?

PS: I also could not make it work when using `fill_in`:

```
     Failure/Error: fill_in '.ts-control', with: value
     
     Capybara::ElementNotFound:
       Unable to find field ".ts-control" that is not disabled within #<Capybara::Node::Element tag="div" path="//HTML[1]/BODY[1]/DIV[1]/DIV[3]/DIV[1]/DIV[1]/DIV[1]/FORM[1]/DIV[1]/FIELDSET[1]/DIV[1]/DIV[1]/DIV[1]">
```

**Comment by filipefurtad0** [comment]
Thank you so much @dacook and @rioug for your early review on this :pray: 
If these two commits are both green then I'll squash them, and carry on with adding tests to this branch, to address #11845.

**Comment by filipefurtad0** [comment]
Hum, sorry - I'm a bit confused:
it seems both `select` (angular?) and `tom-select` (stimulus?) selecting working for the same field (highlighted in the pic below)... Both these approaches seem to work:

`select_tom_select supplier.name, from: 'product_supplier_field'`

`select supplier.name, from: 'product_supplier_id'`

image

Is this expected?

**Comment by rioug** [comment]
> Is this expected? (I'd expect `select` not to work anymore - on the fields marked in green)

`select` is the Capybara one ([doc](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Actions%3Aselect)) :

> #select(value = nil, from: nil, **options) ⇒ [Capybara::Node::Element](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Element)
If from option is present, [#select](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Actions#select-instance_method) finds a select box, or text input with associated datalist, on the page and selects a particular option from it. Otherwise it finds an option inside current scope and selects it. If the select box is a multiple select, [#select](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Actions#select-instance_method) can be called multiple times to select more than one option. The select box can be found via its name, id, [test_id](https://www.rubydoc.info/github/jnicklas/capybara/Capybara#configure-class_method) attribute, or label text. The option can be found by its text.

I am not sure if it's expected to work with "tom select", but if it does I would be inclined to just use it. The git history doesn't really gives us context as to why the tom select helper were added, presumably because `select` didn't work, it's possible that newer version of "tom select" allowed `select` to work.

**Comment by filipefurtad0** [comment]
I have a question about the tom-select dropdowns. After a recent rebase, we seem to be getting several failing specs. I don't know where these changes were introduced... Any ideas? @rioug we've chatted a bit about this in delivery circle, but I'm not sure this is what you meant.

I've tried fixing some of them, like on this [commit](https://github.com/openfoodfoundation/openfoodnetwork/pull/11645/commits/54a25861c4d6e6035e26b3925486d94c5eb54d6e), but it's seems challenging to use the same helper for different drop-downs. Is this expected?

Now I've moved to .spec/admin/order_spec.rb (failing [here](https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/7833846887/job/21375799106?pr=11645).) The naming convention seems to vary bit. Using some existing helpers trigger this error:

```

Capybara::Cuprite::MouseEventFailed: Firing a click at coordinates [481.6937503814697, 331.3999938964844] failed. Cuprite detected another element with CSS selector "html.webkit.chrome.chrome121.linux.js.orientation_landscape.maxw_1280 body.admin.stimulus-reflex-disconnected div.ng-scope div#wrapper div#content.with-sidebar div.container div.row div.twelve.columns div#select-customer fieldset.no-border-bottom div.ts-wrapper.primary.single.plugin-dropdown_input.has-options.focus.input-active.dropdown-active div.ts-dropdown.single.plugin-dropdown_input div.dropdown-input-wrap input.dropdown-input" at this position. It may be overlapping the element you are trying to interact with. If you don't care about overlapping elements, try using node.trigger("click").
from /home/ffurtado/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/cuprite-0.15/lib/capybara/cuprite/page.rb:109:in `rescue in before_click'
Caused by Ferrum::JavaScriptError: Error: click, html.webkit.chrome.chrome121.linux.js.orientation_landscape.maxw_1280 body.admin.stimulus-reflex-disconnected div.ng-scope div#wrapper div#content.with-sidebar div.container div.row div.twelve.columns div#select-customer fieldset.no-border-bottom div.ts-wrapper.primary.single.plugin-dropdown_input.has-options.focus.input-active.dropdown-active div.ts-dropdown.single.plugin-dropdown_input div.dropdown-input-wrap input.dropdown-input, 481.6937503814697, 331.3999938964844
    at Cuprite.mouseEventTest (<anonymous>:296:11)
    at HTMLDivElement.<anonymous> (<anonymous>:1:31)
from /home/ffurtado/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/ferrum-0.14/lib/ferrum/frame/runtime.rb:138:in `handle_error'
```

Indeed, there are overlapping elements (**orders creation page**):

image

image


**Product creation page**:

image

image

I'm unsure how to proceed. Is there a convention we should use:
- to name tom-select dropdowns?
- to use a given helper?

**Comment by filipefurtad0** [comment]
Since we cannot merge this branch and use it during development, I wonder if it is a better idea just to update this branch, shortly before full release (as discussed [here](https://openfoodnetwork.slack.com/archives/C01T75H6G0Z/p1707339007851579), we're looking at about 3 months time). Thoughts?

**Comment by rioug** [comment]
> I have a question about the tom-select dropdowns. After a recent rebase, we seem to be getting several failing specs. I don't know where these changes were introduced... Any ideas? @rioug we've chatted a bit about this in delivery circle, but I'm not sure this is what you meant.
> 

Looks like some of the changes from master got lost, see https://github.com/openfoodfoundation/openfoodnetwork/blob/c0adf947124af8a763c691299acb5bb261536112/spec/support/request/web_helper.rb#L100-L106 , specifically the `send_keys` on line 104. It's missing here : 
https://github.com/filipefurtad0/openfoodnetwork/blob/54a25861c4d6e6035e26b3925486d94c5eb54d6e/spec/support/tom_select_helper.rb#L15-L21

I had a look at the first failing test, this one rspec ./spec/system/admin/order_spec.rb:475 
There are several things going on here, :
* the helper is called with the wrong parameter, it should be : `tomselect_search_and_select customer.email, from: 'customer_search_override'` 
* the helper is missing my fix from master, as shown above.


> I'm unsure how to proceed. Is there a convention we should use:
> - to name tom-select dropdowns?

I believe we just use rails convention here but it does seem a bit inconsistent

`tomselect_multiselect`, `tomselect_search_and_select`, `tomselect_select` all expect the name of the select component, ie: 

tomselect_name

`select_tom_select` on the other hand expect an html id, and will go look for tomselect in the component's children.

> - to use a given helper?

I don't know the history of these helpers, but it looks like `tomselect_search_and_select` and `select_tom_select` are doing the same thing. So I would suggest to keep only one of these two and keeping the naming consistent. I added `tomselect_select` because sometimes tomselect is setup as a normal select, ie you see the options dropdown is already populated when you click on tomselect. Even though the search functionality is present, it not necessary to search to find your your option. In this scenario `tomselect_search_and_select` was breaking due to some race condition. See:  https://github.com/openfoodfoundation/openfoodnetwork/commit/98545741e6069a2ec288552bac7de6e4e690a450
So I would say use: 
* `tomselect_search_and_select` when searching is actually required, ie: the order spec example above.
* `tomselect_select` when searching is not required
I am not sure about `tomselect_multiselect` though.

Hopefully that should unblock you !

**Comment by filipefurtad0** [comment]
Thanks @rioug !

> Looks like some of the changes from master got lost, see 

Indeed. This might have been to a bad rebase (on me :see_no_evil: )
At some point in the discussion above we've discussed moving helpers to their own file. I've now added that change on the respective helper which was indeed lost. 

> the helper is called with the wrong parameter, it should be : tomselect_search_and_select customer.email, from: 'customer_search_override'
    the helper is missing my fix from master, as shown above.

Having this done, brings the build back to green - which is great. I was a bit lost here.


>    tomselect_search_and_select when searching is actually required, ie: the order spec example above.
    tomselect_select when searching is not required
    I am not sure about tomselect_multiselect though.

Agree to remove `select_tom_select` if we don't need it, but the thing is, I can't seem to make any of the three above helpers work for the Unit Size dropdown, all search for the `name` attribute, and there is none defined within `select` element:


image


Hence, I've created this additional helper the `select_tom_select`, which - as you say looks for the `id:`:

image

I guess we can only get rid of `select_tom_select`, once we add a name attribute to that `select` element. Not sure I'm missing something - what do you think?

**Comment by dacook** [comment]
Hi Filipe, I've updated the `tomselect_*select` methods in this PR to be more flexible:
https://github.com/openfoodfoundation/openfoodnetwork/pull/12183/commits/520482b8b69ca5e5de618235bce80df9bbabd998

Once it's merged, you'll have the option of using the label text (just like `fill_in`), which I think is generally better for system specs.
Eg: 
```ruby
        tomselect_select "Weight (g)", from: "Unit Size"
```

**Comment by dacook** [comment]
Hi @filipefurtad0 , I went ahead and rebased this to prepare for merge. I condensed some of the past commits, and added some more fixes. 
I also added some "TODO" comments where we are still referring to the old Bulk Edit Products screen. I'd like to migrate those over to the new screen soon, but I think it can wait so that we can merge this long-running branch.

**Review by dacook — APPROVED** [review]
Although the latest changes are mine, I'm approving to resolve the "changes requested".

File: spec/system/admin/product_import_spec.rb
Comment (rioug) [inline spec/system/admin/product_import_spec.rb:99]: Shouldn't that be tracked in an issue ?

**Review by rioug — APPROVED** [review]
Looks good !

File: spec/system/admin/product_import_spec.rb
Comment (dacook) [reply spec/system/admin/product_import_spec.rb:99]: Yes, I'll add checkboxes in the above issue 👍

Comment (dacook) [reply spec/system/admin/product_import_spec.rb:99]: Done

File: spec/system/admin/reports_spec.rb
Comment (mkllnk) [inline spec/system/admin/reports_spec.rb]: This looks like a mistake.

Comment (mkllnk) [reply spec/system/admin/reports_spec.rb]: But it seems to have been fixed later.

File: spec/system/admin/bulk_product_update_spec.rb
Comment (mkllnk) [inline spec/system/admin/bulk_product_update_spec.rb:33]: The main commit message doesn't reflect the commit content.

File: spec/system/admin/order_cycles/complex_editing_multiple_updation_spec.rb
Comment (mkllnk) [inline spec/system/admin/order_cycles/complex_editing_multiple_updation_spec.rb:39]: Or we could setup the enterprise properly to avoid the warning?

File: spec/system/admin/order_cycles/simple_spec.rb
Comment (mkllnk) [inline spec/system/admin/order_cycles/simple_spec.rb:524]: Can we dismiss the warning first? Otherwise this is a UX issue. Users can't "find button and trigger save" with their mouse.

File: spec/system/admin/order_spec.rb
Comment (mkllnk) [inline spec/system/admin/order_spec.rb]: Another conflict that could have been squashed.

File: spec/system/admin/products_v3/products_spec.rb
Comment (mkllnk) [inline spec/system/admin/products_v3/products_spec.rb]: I think that the previous selectors were to avoid false passes of the spec. Imagine this spec running with _Enterprise 4_. It would show the number 4 even without displaying that stock.

**Review by mkllnk — APPROVED** [review]
Huge chunk of work. Well done for taming this.

There are many little bits I would have like to see improved in the history but this needs to be merged as soon as possible to avoid more rebase errors.

File: spec/system/admin/reports_spec.rb
Comment (filipefurtad0) [reply spec/system/admin/reports_spec.rb]: Yeah... Sorry. Indeed, this might have happened a couple of times. This branch lived for quite a long time and was fixed for conflicts/rebased countess times. I'm not sure if it would not have been a better approach to open this branch, at a later stage. Maybe a topic for discussion if we get to do a retro on BUU.

File: spec/system/admin/order_cycles/complex_editing_multiple_updation_spec.rb
Comment (filipefurtad0) [reply spec/system/admin/order_cycles/complex_editing_multiple_updation_spec.rb:39]: Done in #12652 :pray:

File: spec/system/admin/products_v3/products_spec.rb
Comment (filipefurtad0) [reply spec/system/admin/products_v3/products_spec.rb]: Do you think we should revert to the xpath code? Maybe an alternative would be to include the count: attribute, or make the numbers a bit more specific, like 111, 222, 333 - I know, in theory, we could still have Enterprise 111, but maybe more unlikely.

What do you think @mkllnk @dacook ?

Comment (mkllnk) [reply spec/system/admin/products_v3/products_spec.rb]: Yes, I would go back to the xpath. If this was a more common use case then I would create helpers for that. Aren't there some helpers for accessing table rows by a displayed name, for example? That would be better.

Comment (dacook) [reply spec/system/admin/products_v3/products_spec.rb]: Yes, in fact the inputs have labels associated with them, so something this should work:
```
within row_containing_name("variant name") do
  expect(page).to have_field "On demand", checked: true
  expect(page).to_not have_field "On Hand", with: 20
end
```
(hmm, we need to fix that inconsistent capitalisation..)

Comment (dacook) [reply spec/system/admin/products_v3/products_spec.rb]: Sorry, that was targeting the fields inside the popout.

If we want to check the display, we can target the button (which opens the popout):
```
within row_containing_name("variant name") do
  expect(page).to have_button "On Hand", with: "On demand"
  expect(page).to_not have_button "On Hand", with: 20
end
```

Comment (dacook) [reply spec/system/admin/products_v3/products_spec.rb]: Even better might be to say something like this. But as far as I know Capybara doesn't have a convenient way to check all text (including button labels). We could make a helper for that, but it's not worth the bother..

```
within row_containing_name("variant name") do
  within column("On Hand") do
    expect(page).to have_text "On demand"
    expect(page).to_not have_text "20"
  end
end
```

**Pull request merged**
